[WIP] Add XBlocks

symbolist opened the pull request.

bradenmacdonald commented on labxchange_xblocks/case_study_block.py:
```diff
@@ -0,0 +1,66 @@
+from xblock.core import XBlock
+from xblock.fields import Scope, String
+from xblockutils.studio_editable import (
+    StudioContainerWithNestedXBlocksMixin,
```
FYI, the UI part of `StudioContainerWithNestedXBlocksMixin` won't [yet?] be working in the new runtime, but there is a Content Libraries REST API endpoint that the LabXchange backend can call to add a specific XBlock type as a child of an existing block. Which might be what you'd prefer anyways?

symbolist replied:
Yes, we will have the custom UI in LX. This is for testing in studio and developing the initial content.

bradenmacdonald commented on labxchange_xblocks/simulation_block.py:
```diff
@@ -0,0 +1,49 @@
+from .utils import AuthorViewBlockMixin, StudentViewBlockMixin, _
+
+
+class SimulationBlock(XBlock, AuthorViewBlockMixin, StudioEditableXBlockMixin, StudentViewBlockMixin):
```
The simulation XBlock needs to store a JSON object as user state, right? And it needs a handler that the simulation HTML5 code can use to get/set that state?

symbolist replied:
It's not needed right now so will add it next quarter.

bradenmacdonald commented:
There are some existing XBlock conventions that I'd like you to use here please:

* Any data that defines how to render the XBlock but is not user-specific should be in a method called:
    ```python
    def student_view_data(self, context=None):
    ```
    ([dndv2 example](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/0653987bc4ee67809bc95a433f4f17dc7aedf732/drag_and_drop_v2/drag_and_drop_v2.py#L307-L360), [video example](https://github.com/edx/edx-platform/blob/8dc4fd1a2d8c79776d6b452f45cdffee3264f993/common/lib/xmodule/xmodule/video_module/video_module.py#L1058-L1136))
    * Such data gets cached aggressively in memcached and can be retrieved using the [Course Blocks API](https://openedx.atlassian.net/wiki/spaces/MS/pages/29688043/Course+Blocks+API) (for modulestore) and TBD for blockstore. In the meantime, it makes sense to have an inherited `v1_student_view_data` handler that can retrieve this data, but the underlying methods should follow the existing API as documented in the edX wiki and implemented in existing blocks. However, I can probably add a method to get this data to the new runtime quite easily, in which case we won't need the `v1_student_view_data` handler at all.
    * So please rename `student_view_context` to `student_view_data`
* Any data that _is_ user-specific (such as the assignment scores) must not be in the `student_view_data` since that is aggressively cached in a user-agnostic way. Instead, convention is for such data to be in an XBlock handler called `student_view_user_state` - see [this drag and drop v2 example](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/0653987bc4ee67809bc95a433f4f17dc7aedf732/drag_and_drop_v2/drag_and_drop_v2.py#L644-L649)

bradenmacdonald commented on labxchange_xblocks/utils.py:
```diff
@@ -0,0 +1,98 @@
+
+class AuthorViewBlockMixin(XBlockMixin):
+    """
+    Mixin which sets the student_view() as the author_view().
```
Is this necessary? What is the effect of leaving this mixin out? I thought studio would use the student_view anyways.

symbolist commented:
@bradenmacdonald I have made the following changes:

1. Renamed `student_view_context()` to `student_view_data()`.
2. Had to keep a separate handler though since the API signature is different.
2. Moved user state for Assignment out to a separate handler: `student_view_user_state`.

bradenmacdonald commented:
@symbolist Code looks good now, thanks! I will test this together with the other PR and then approve.

bradenmacdonald commented:
:+1: for what you have so far.

- [x] I tested this: via the LabXchange PR "XBlock custom learner views"
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation: the existing readme has an overview

bradenmacdonald commented on setup.py:
```diff
@@ -85,4 +102,14 @@ def is_requirement(line):
     ],
+    entry_points={
+        'xblock.v1': [
+            'assignment = labxchange_xblocks.assignment_block:AssignmentBlock',
```
Let's' prefix these with `lx_`

ormsbee commented on labxchange_xblocks/templates/story_in_science_student_view.html:
```diff
@@ -0,0 +1,13 @@
+        </div>
+    </div>
+    <div class="story-in-science-block-story">
+        {{ story|safe }}
```
Is it necessary to put the `safe` filter here? I get why you'd do it for the case of the container blocks that echo our their childrens' HTML, but shouldn't HTML at this level be escaped?

symbolist replied:
I double checked this. Without `safe`, the string is [escaped](https://docs.djangoproject.com/en/2.2/ref/templates/builtins/#escape) which means all html tags are rendered as text instead of markup.

Is there a standard way of escaping javascript only?

ormsbee replied:
Not that I'm aware of. I mean, you could use `bleach`, but that had some serious performance issues the last time I checked (we use it somewhere in edx-platform, I forget where exactly).

bradenmacdonald replied:
With the new runtime, all XBlocks get rendered in a sandboxed iframe (or at least have the ability to be so rendered, depending on the host application), so `bleach` isn't necessary. Even if user-input JS code gets run, it has no access to the LMS's domain so can't access anything.

ormsbee commented on requirements/base.txt:
```diff
@@ -9,8 +9,8 @@
-django==1.11.23
-fs==2.4.10                # via xblock
+django==1.11.24
+fs==2.4.11                # via xblock
```
Please use version ranges instead of pinning things to exact versions like this, to make upgrades easier.

symbolist replied:
This is using the setup from https://github.com/edx/cookiecutter-django-app:

1. The unpinned dependencies are in the `.in` files.
2. `make upgrade` using `pip-compile` to pin versions in `.txt` files.
3. `setup.py` has `install_requires=load_requirements('requirements/base.in'),`

So it should work, right or am I missing something?

ormsbee replied:
Ah, I see. I had thought that the `.in` thing was something we used for the deployed (e.g. at the service level), and that XBlocks themselves were unpinned as much as possible. But yeah, I guess it will work the way you described.

symbolist commented:
@ormsbee I have sent you an invite to a course with these XBlocks setup. Let me know whenever you signup and I can add you as course staff.

ormsbee commented:
@symbolist: Did the signup, username dave

symbolist commented:
@ormsbee Added to course team.

One note: there is very little styling included. Adding it via the LX frontend to avoid need to bump up this package version whenever we tweak anything.

symbolist commented:
Hmm. Looks like the import did not work correctly.

symbolist commented:
@ormsbee  Okay, so I was in the middle of a rebase when I exported and so the archive was missing the olx. I have created another course with the content and invited you as course staff.

Sorry for the inconvenience!

symbolist commented:
@ormsbee Let me know whenever you have completed your testing and I can merge this.

ormsbee commented:
@symbolist: Simulation box size seems a bit small, but that's out of the scope of my concerns. LGTM in terms of merging. Thank you.
